src/i_oalsound.c

```diff
@@ -352,6 +362,78 @@ const char **I_OAL_GetResamplerStrings(void)
     return strings;
 }
 
+static void SetEqualizer(void)
+{
+    ALuint uiEffectSlot, uiEffect, uiFilter;
+    ALCint iSends = 0;
+    int gain_levels_pct[4] = {0};
+    float gain_levels[4] = {0.f};
+
+    if (!oal || !oal->EXT_EFX)
+    {
+        return;
+    }
+
+    // Check the actual number of Auxiliary Sends available on each Source.
+
+    alcGetIntegerv(oal->device, ALC_MAX_AUXILIARY_SENDS, 1, &iSends);
+
+    if (iSends < 1)
+    {
+        return;
+    }
+
+    if (strcmp(snd_equalizer, EQUALIZER_DEFAULT) == 0 ||
+        sscanf(snd_equalizer, "%d %d %d %d",
+               &gain_levels_pct[0], &gain_levels_pct[1],
+               &gain_levels_pct[2], &gain_levels_pct[3]) != 4)
+    {
+        return;
+    }
+
+    alGenAuxiliaryEffectSlots = FUNCTION_CAST(LPALGENAUXILIARYEFFECTSLOTS, alGetProcAddress("alGenAuxiliaryEffectSlots"));
+    alGenEffects = FUNCTION_CAST(LPALGENEFFECTS, alGetProcAddress("alGenEffects"));
+    alEffecti = FUNCTION_CAST(LPALEFFECTI, alGetProcAddress("alEffecti"));
+    alEffectf = FUNCTION_CAST(LPALEFFECTF, alGetProcAddress("alEffectf"));
+    alAuxiliaryEffectSloti = FUNCTION_CAST(LPALAUXILIARYEFFECTSLOTI, alGetProcAddress("alAuxiliaryEffectSloti"));
+    alGenFilters = FUNCTION_CAST(LPALGENFILTERS, alGetProcAddress("alGenFilters"));
+
+    if (!alGenAuxiliaryEffectSlots ||
+        !alGenEffects ||
+        !alEffecti ||
+        !alEffectf ||
+        !alAuxiliaryEffectSloti ||
+        !alGenFilters)
+    {
+        return;
+    }
+
+    alGenAuxiliaryEffectSlots(1, &uiEffectSlot);
+    alGenEffects(1, &uiEffect);
+    alEffecti(uiEffect, AL_EFFECT_TYPE, AL_EFFECT_EQUALIZER);
+
+    // Gains vary from 0.126 up to 7.943, which means from -18dB attenuation
+    // up to +18dB amplification, i.e. 20*log10(gain).
+
+    for (int i = 0; i < 4; i++)
+    {
+        gain_levels[i] = BETWEEN(0.126, 7.943, gain_levels_pct[i]/100.f);
+    }
+
+    alEffectf(uiEffect, AL_EQUALIZER_LOW_GAIN, gain_levels[0]);
+    alEffectf(uiEffect, AL_EQUALIZER_MID1_GAIN, gain_levels[1]);
+    alEffectf(uiEffect, AL_EQUALIZER_MID2_GAIN, gain_levels[2]);
+    alEffectf(uiEffect, AL_EQUALIZER_HIGH_GAIN, gain_levels[3]);
+
+    alAuxiliaryEffectSloti(uiEffectSlot, AL_EFFECTSLOT_EFFECT, uiEffect);
+    alGenFilters(1, &uiFilter);
+
+    for (int i = 0; i < MAX_CHANNELS; i++)
+    {
+        alSource3i(oal->sources[i], AL_AUXILIARY_SEND_FILTER, uiEffectSlot, 0, uiFilter);
```

Comment:
I suggest unloading the effects if the equalizer is "flat" to minimize overhead. Maybe also call the default "off" to avoid confusion.

Reply:
You mean with `AL_FILTER_NULL`?